src/CPUActivityAgent.cpp

```diff
@@ -0,0 +1,532 @@
+            child_pol = in_policy;
+        }
+    }
+
+    // Indicate whether to send the policy down to children
+    bool CPUActivityAgent::do_send_policy(void) const
+    {
+        return true;
+    }
+
+    void CPUActivityAgent::aggregate_sample(const std::vector<std::vector<double> > &in_sample,
+                                        std::vector<double>& out_sample)
+    {
+
+    }
+
+    // Indicate whether to send samples up to the parent
+    bool CPUActivityAgent::do_send_sample(void) const
+    {
+        return false;
+    }
+
+    void CPUActivityAgent::adjust_platform(const std::vector<double>& in_policy)
+    {
+        m_do_write_batch = false;
+
+        if (m_update_qm_max_rate) {
+            for (auto it = in_policy.begin() + M_POLICY_FIRST_UNCORE_FREQ;
+                 it != in_policy.end() && std::next(it) != in_policy.end();
+                 std::advance(it, 2)) {
+
+                if (!std::isnan(*it)) {
+                    auto uncore_freq = static_cast<uint64_t>(*it);
+                    auto max_mem_bw = *(it + 1);
+                    // Not valid to have NAN max mem bw for uncore freq.
+                    GEOPM_DEBUG_ASSERT(!std::isnan(max_mem_bw),
+                                       "mapped CPU_UNCORE_FREQUENCY with no max memory bandwidth assigned.");
+                    m_qm_max_rate[uncore_freq] = max_mem_bw;
+                }
+            }
+            // Warn the user if they have not entered a policy with memory bandwidth characterization.
+            // In this case the maximum uncore frequency will be used.  This can easily be converted
+            // to an assertion if it should be considered a failure case.
+            if (m_qm_max_rate.size() == 0) {
+                std::cerr << "Warning: <geopm> CPUActivityAgent did not receive a policy containing memory " <<
+                             "bandwidth characterization.  This may negatively impact agent performance." << std::endl;
+            }
+            m_update_qm_max_rate = false;
+        }
+
+        // Per package freq
+        std::vector<double> uncore_freq_request;
+        m_resolved_f_uncore_efficient = in_policy[M_POLICY_UNCORE_FREQ_EFFICIENT];
+        m_resolved_f_uncore_max = in_policy[M_POLICY_UNCORE_FREQ_MAX];
+        double f_uncore_range = in_policy[M_POLICY_UNCORE_FREQ_MAX] - in_policy[M_POLICY_UNCORE_FREQ_EFFICIENT];
+
+        //m_resolved_f_uncore_efficient = 0;
+        for (int domain_idx = 0; domain_idx < M_NUM_PACKAGE; ++domain_idx) {
+            double uncore_freq = (double) m_uncore_freq_status.at(domain_idx).value;
+
+            /////////////////////////////////////////////
+            // L3 Total External Bandwidth Measurement //
+            /////////////////////////////////////////////
+            auto qm_max_itr = m_qm_max_rate.lower_bound(uncore_freq);
+            if(qm_max_itr != m_qm_max_rate.begin()) {
+                qm_max_itr = std::prev(qm_max_itr, 1);
+            }
```

Comment:
Please comment on what's going on here.

You're assign ```qm_max_itr``` an iterator that 1 *before* the lower bound of the current uncore frequency?  Why?

Reply:
This is intended to handle cases where the signal does not cleanly map to entries in the qm_max_rate keys (or vice versa).  
For a map of {1.9: a, 2.0: b, 2.2: c, 2.3: d} lower_bound(2.2) & lower_bound(2.1) will both return an iterator to (2.2: c).  Similarly lower_bound(2.21) will return an iterator to (2.3: d).  
As a simplification to address this issue while targeting energy efficiency we map to one iterator below the lower bound.

Alternative solutions that may be implemented
- Only use std::prev if the uncore_freq doesn't match a key in the map
- Round (or floor) uncore_freq to a key in the m_qm_max_rate map before doing lower_bound
- Require map entries be valid values for the hardware specified by checking them vs the step size (not currently possible with existing GEOPM signals)

Reply:
"Only use std::prev if the uncore_freq doesn't match a key in the map" has been implemented in the latest set of changes.  Performance testing against 4 workloads of interest has shown no harm or better energy savings & perf deg for each workload.